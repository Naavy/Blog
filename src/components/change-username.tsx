import React from 'react';
import { TextField, Button, Box } from '@material-ui/core';
import { connect } from 'react-redux';
import { AppState, User } from '../store/types';
import { Dispatch } from 'redux';
import { changeUserName } from '../store/actions';

interface State {
  lastName: string;
  firstName: string;
}

interface StoreProps {
  selectedUser?: User;
}

interface DispatchProps {
  changeUserName: (id: string, userData: Record<string,string>) => any;
}

type Props = StoreProps & DispatchProps;

export class ChangeUserName extends React.Component<Props> {
  readonly state: State = {
    lastName: '',
    firstName: ''
  }

  setUserData = () => {
    if(this.props.selectedUser) {
      this.setState({
        lastName: this.props.selectedUser.last_name,
        firstName: this.props.selectedUser.first_name
      })
    }
  }

  componentDidMount() {
    this.setUserData();
  }

  componentDidUpdate(prevProps: any) {
    if(prevProps.selectedUser !== this.props.selectedUser) {
      this.setUserData()
    }
  }

  sendForm(event: React.FormEvent<HTMLFormElement>) {
    event.preventDefault();
    if(this.props.selectedUser) {
      this.props.changeUserName(
        this.props.selectedUser.id,
        { last_name: this.state.lastName, first_name: this.state.firstName }
      );
    }
  }

  render() {
    return (
      <React.Fragment>
        {this.props.selectedUser &&
        <form onSubmit={(event) => this.sendForm(event)}>
          <Box style={{ display: "flex", marginBottom: "30px" }}>
            <TextField
              label="Last Name"
              value={this.state.lastName}
              style={{ marginRight: '20px' }}
              onChange={(event) => this.setState({ lastName: event.target.value})}
            />
            <TextField
              label="First Name"
              value={this.state.firstName}
              onChange={(event) => this.setState({ firstName: event.target.value})}
            />
          </Box>
          <Button
            variant="contained"
            type="submit"
            style={{ width: '25%' }}
          >
            Change
          </Button>
        </form>
        }
      </React.Fragment>
    )
  }
}

const mapStoreToProps = (state: AppState): StoreProps => {
  return {
    selectedUser: state.selectedUser,
  }
}

const mapDispatchToProps = (dispatch: Dispatch): DispatchProps => {
  return {
    changeUserName: (id: string, userData: Record<string,string>) => {
      return dispatch(changeUserName(id, userData))
    }
  }
}

export default connect(mapStoreToProps, mapDispatchToProps)(ChangeUserName);
