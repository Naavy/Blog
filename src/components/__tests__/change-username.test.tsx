import React from 'react';
import { shallow } from 'enzyme';
import { ChangeUserName } from '../change-username';

const store = {
    lastName: 'TestLastName',
    firstName: 'TestFirstName',
    selectedUser: {
      "id": "1",
      "first_name": "FirstNameTest",
      "last_name": "LastNameTest",
      "email": "mail@test.com",
    },
    changeUserName: jest.fn
  }


describe('ChangeUserName', () => {
  it('should render', () => {
    // when
    const wrapper = shallow(<ChangeUserName {...store} />);

    // then
    expect(wrapper).toMatchSnapshot();
  });

  it('should call changeUserName, when click "change" ', () => {
    // given
    const mockedChangeUserName = jest.fn().mockResolvedValue(null);
    const testStore = {
        ...store,
        changeUserName: mockedChangeUserName
    }

    const wrapper = shallow(<ChangeUserName {...testStore} />);

    // when
    wrapper.find('form').simulate('submit', { preventDefault: () => null });

    // then
    expect(mockedChangeUserName).toHaveBeenCalled();
  });
});