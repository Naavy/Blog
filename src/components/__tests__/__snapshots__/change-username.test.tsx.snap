// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangeUserName should render 1`] = `ShallowWrapper {}`;
